src/Data/HashMap/Linear.hs

```diff
@@ -1,73 +1,239 @@
+      v #-> (HashMap k v, RobinQuery k) #-> HashMap k v
+    insertFromQuery v (HashMap (size,count) arr, IndexToInsert (k,psl) ix) =
+      HashMap (size,count+1) (write arr ix (k,v,psl))
+    insertFromQuery v (HashMap sizes arr, IndexToUpdate (k,psl) ix) =
+      HashMap sizes (write arr ix (k,v,psl))
+    insertFromQuery v (HashMap sizes arr, IndexToSwap (k,psl) ix) =
+      (Unsafe.toLinear swapFromRead) (read arr ix) sizes ix (k,v,psl)
+
+    -- XXX: This re-does the probe sequence of (k',v')
+    swapFromRead :: EqHashable k => (RobinArr k v, RobinVal k v) ->
+      (Int, Int) -> Int -> RobinVal k v #-> HashMap k v
+    swapFromRead (arr', (k',v',_)) sizes ix (k,v,psl) =
+      insert (HashMap sizes (write arr' ix (k,v,psl))) k' v'
 
 -- | If present, deletes key-value pair, otherwise does nothing
-delete :: Hashable k => HashMap k v #-> k -> HashMap k v
-delete hmap k = alter (\_ -> Nothing) k hmap
-
+delete :: EqHashable k => HashMap k v #-> k -> HashMap k v
+delete hmap k = deleteFromQuery $ queryIndex hmap k
+  where
+    deleteFromQuery :: EqHashable k =>
+      (HashMap k v, RobinQuery k) #-> HashMap k v
+    deleteFromQuery (h, IndexToInsert _ _) = h
+    deleteFromQuery (h, IndexToSwap _ _) = h
+    deleteFromQuery (h, IndexToUpdate _ ix) =
+      shiftBack h ix
+
+    -- Continue to shift back the next element and decrease the PSL
+    -- until we see an element with PSL 0. The arguments are:
+    -- `deleteShiftFrom (hashmap, size) index`
+    shiftBack :: EqHashable k => HashMap k v #-> Int -> HashMap k v
+    shiftBack (HashMap (size,count) arr) ix =
+      HashMap (size,count-1) ((Unsafe.toLinear shiftBackArr) arr size ix)
+
+    -- XXX: make this linear
+    shiftBackArr :: EqHashable k =>
+      RobinArr k v -> Int -> Int -> RobinArr k v
+    shiftBackArr arr size ix =
+      case read arr (ix + 1 `mod` size) of
+        (arr', (_,_,PSL p)) | p == 0 ->
+          case read arr' ix of
+            (arr'', (k',v',_)) -> write arr'' ix (k',v',-1)
+        (arr', (k,v,PSL p)) ->
+          shiftBackArr (write arr' ix (k,v,PSL (p-1))) size (ix+1 `mod` size)
+
+-- | If the load is too high resize by tripling the array.
+maybeResize :: EqHashable k => HashMap k v #-> HashMap k v
+maybeResize (HashMap (size, count) arr)
+   | count*3 < size = HashMap (size, count) arr
+   | otherwise = HashMap (size*3, count) (Unsafe.toLinear tripleArr arr)
+  where
+    tripleArr :: EqHashable k => RobinArr k v -> RobinArr k v
+    tripleArr arr = case read arr 0 of
+      (arr', (k,v,_)) ->
+        moveHanging size (resizeSeed (size*3) (k,v,-1) arr', 0)
+
+    -- | After copying over to an array trice the original size,
+    -- we need to move all "hanging elements", i.e., those triples (k,v,psl)
+    -- where the psl is greater than the index of the cell.
+    -- Theorem. The hanging elements of a robin array form a prefix.
+    -- The proof is left as an exercise to the reader.
+    -- Arguments: `adjustModedPrefix original-size (arr,ix)`
```

Comment:
This is a typo. The arguments line should use `moveHanging` and not `adjustModedPrefix`